lib/Dialect/ESI/cosim/cosim_dpi_server/Endpoint.cpp

```diff
@@ -0,0 +1,75 @@
+//===- EndPoint.cpp - Definitions for EndPointRegistry ----------*- C++ -*-===//
+//
+// Definitions for Cosim EndPoint and EndPointRegistry.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Endpoint.h"
+
+using namespace circt::esi::cosim;
+
+Endpoint::Endpoint(uint64_t sendTypeId, int sendTypeMaxSize,
+                   uint64_t recvTypeId, int recvTypeMaxSize)
+    : sendTypeId(sendTypeId), recvTypeId(recvTypeId), inUse(false) {}
+Endpoint::~Endpoint() {}
+
+uint64_t Endpoint::getSendTypeId() const { return sendTypeId; }
+uint64_t Endpoint::getRecvTypeId() const { return recvTypeId; }
+
+bool Endpoint::setInUse() {
+  Lock g(m);
+  if (inUse)
+    return false;
+  inUse = true;
+  return true;
+}
+
+void Endpoint::returnForUse() {
+  Lock g(m);
+  if (!inUse)
+    throw std::runtime_error("Cannot return something not in use!");
+  inUse = false;
+}
+
+void EndpointRegistry::registerEndpoint(int epId, uint64_t sendTypeId,
+                                        int sendTypeMaxSize,
+                                        uint64_t recvTypeId,
+                                        int recvTypeMaxSize) {
+  Lock g(m);
+  if (endpoints.find(epId) != endpoints.end())
+    throw std::runtime_error("Endpoint ID already exists!");
+  // The following ugliness adds an Endpoint to the map of Endpoints. The
+  // Endpoint class has its copy constructor deleted, thus the metaprogramming.
+  endpoints.emplace(std::piecewise_construct,
+                    // Map key.
+                    std::forward_as_tuple(epId),
+                    // Endpoint constructor args.
+                    std::forward_as_tuple(sendTypeId, sendTypeMaxSize,
```

Comment:
Were you able to simplify this?  I'm not sure why you need forwarding for trivial types like this.

Reply:
It's constructing a non-trivial type -- the second `forward_as_tuple` are constructor arguments to the `Endpoint` class which has its copy constructor deleted.

Reply:
I've spent multiple hours trying but my metaprogramming skills are weak.